### ISSUE TITLE
Don't send all credstore entries to demon on docker build

### ISSUE THREAD

**Issue by jeanlaurent** [docker/docker#31517]
In Docker For Mac/Win in 17.03.0-ce we make default the usage of `docker-credential-helper`.

A user reported that we scan each and every entry in the all user keychains, as per https://github.com/docker/for-mac/issues/1371.

This behaviour is observable, only on `docker build` and not `docker pull`.

This leads us to believe that https://github.com/docker/docker/blob/master/cli/command/image/build.go#L280 is probably partly  faulty. Along with https://github.com/docker/docker-credential-helpers/blob/master/osxkeychain/osxkeychain_darwin.c#L118 which seems to open all keychains.
All entries in the keychains are sent to the daemon. Which in turn triggered the behaviour described by the user.

They should probably be filtered by the one used by `docker login`.

**Output of `docker version`:**

```
> docker version
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.12-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 5.572 GiB
Name: moby
ID: T4H4:FHZA:GOQW:7GND:BENW:HYGI:JHRC:H3XI:OQQP:YW22:6YNH:O2DX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 16
 Goroutines: 26
 System Time: 2017-03-03T13:24:29.210944844Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by jeanlaurent** [comment]
cc @vdemeester

**Comment by vdemeester** [comment]
> They should probably be filtered by the one used by `docker login`.

i.e. those that are in the `config.json` in the following form.

```
{
        "auths": {
                "https://index.docker.io/v1/": {},
                "localhost:5000": {},
                "quay.io": {}
        }
}
```

cc @docker/security-team

**Comment by n4ss** [comment]
We could tag keychain entries with what's in `config.json:auth` when creating them with:

- "docker auths" for example in  [kSecAttrDescription](https://developer.apple.com/reference/security/ksecattrdescription)

- serialized associated data in [kSecAttrLabel](https://developer.apple.com/reference/security/ksecattrlabel)

That way, entry retrieval wouldn't go through all keychains entries.

**Comment by ebriney** [comment]
@n4ss : that's ok for osxkeychain but what about `wincred` and `secretservice`?
And if we use `auths` urls, we don't even need to list the credStore keys.
If the goal is to remove those urls from `config.json` that is another thing...

**Comment by dlanger** [comment]
I'm now seeing behaviour where credentials aren't being used on `docker build`, but are on `docker pull`:

```
$ make start
build     │ Sending build context to Docker daemon 503.8 kB
build     │ Step 1/10 : FROM my_private_ecr/image_name:tag_name
build     │ Pulling repository my_private_ecr/image:name
build     │ unauthorized: authentication required
ERROR: build error: exit status 1

$ docker pull my_private_ecr/image_name:tag_name
tag_name: Pulling from image_name
b7f33cc0b48e: Already exists
```
`make start` is a wrapper script that does some unrelated things, eventually getting down to a `docker build`.

I'm on the latest Docker for Mac (15654).
```
docker --version
Docker version 17.03.0-ce, build 60ccb22
```

**Comment by n4ss** [comment]
> that's ok for osxkeychain but what about wincred and secretservice?

For Windows,  a PR can be submitted to`winsec` to add attributes to credentials when storing and also to lookup by attribute. Windows handles that: [Credentials](https://msdn.microsoft.com/en-us/library/windows/desktop/aa374788(v=vs.85).aspx) and their [Attributes](https://msdn.microsoft.com/en-us/library/windows/desktop/aa374790(v=vs.85).aspx)

For Linux,it seems  we already use `libsecret` secrets' attributes. We can use that to filter the lookup if needed.

**Comment by vdemeester** [comment]
@n4ss I think that could/should work 👍

**Comment by n4ss** [comment]
I can write the patch by the way.

**Comment by n4ss** [comment]
I'm following-up here for the changes to be made on `docker-credentials-helper`: :https://github.com/docker/docker-credential-helpers/issues/46

**Comment by KingEmet** [comment]
@jeanlaurent I'm the person that made the credential stores/helpers themselves the source of truth (as opposed to the docker config) when it comes to enumerating stored credentials ( #26354 ). Delivering the entire universe of credentials to the daemon during `docker build`s was existing behavior which was potentially broken by credential helpers. The 'correct' fix in this case would probably be to limit the credentials passed to the daemon.

**Comment by n4ss** [comment]
@KingEmet my ongoing patch is currently done in the helpers but I agree with you. The best way long-term should probably be to filter on daemon creds requests and helpers should themselves namspace/tag the docker creds entries as such.

**Comment by justincormack** [comment]
Moved to P0 after discussion with @n4ss.

**Comment by justincormack** [comment]
cc @icecrime ^^

**Comment by simonferquel** [comment]
After discussions, we found we couldn't fix this issue elegantly / easily without introducing changes in the communication between credential helpers and docker client.

One interesting way to solve that permanently (and much more elegantly), is to leverage @tonistiigi work on streamed context (https://github.com/docker/docker/issues/31829) once it is done, to expose grpc methods allowing to request credentials on demand from the builder to the client in addition to the filesystem context.

**Comment by vdemeester** [comment]
Removed the P0 as this is fixed for desktop version — the docker-credential-x binaries now use labels to just get/ask for the docker credentials and not all of them 👼

**Comment by jimschubert** [comment]
@dlanger I found today that current docker stable and edge have issues with logging out of an internal registry on mac (I assume it's one that was logged into with a previous docker version).  This resulted in all `docker build` commands failing to pull base images for me if the base image didn't already exist locally. #33252 discusses this issue specifically.

When you `docker logout internal_registry`, you can open keychain access and under login/passwords you may still see "Docker Credentials". Delete that password, log back into your internal registry, and it may resolve your issue.

**Comment by goffinf** [comment]
The same behaviour described above also effects other platforms. I am running on AWS (RHEL 7.3) with docker version 17.05.

The private registry in this case is ECR and is confgiured thus in .docker/config.json -

```
{
    "credsStore": "ecr-login"
}
```

docker push and pull work successfully. You can use docker login beforehand per the AWS docs here -

https://aws.amazon.com/blogs/compute/authenticating-amazon-ecr-repositories-for-docker-cli-with-credential-helper/

```
$ aws ecr get-login
docker login –u AWS –p password –e none https://aws_account_id.dkr.ecr.us-east-1.amazonaws.com
```

... or (as we do and the same doc goes on to suggest) engage the ECR Credentials Helper which does it for you.

The behaviours are ... (lets say I have a copy on the alpine:3.5 image in a private repo)

docker pull 123456789012.dkr.ecr.eu-west-1.amazonaws.com/alpine:3.5  (works)
docker push 123456789012.dkr.ecr.eu-west-1.amazonaws.com/alpine:3.5  (works)

But build FAILS

```
Sending build context to Docker daemon  4.096kB
Step 1/5 : FROM 123456789012.dkr.ecr.eu-west-1.amazonaws.com/alpine:3.5
Pulling repository 300820918606.dkr.ecr.eu-west-1.amazonaws.com/alpine
unauthorized: authentication required
```

If the image is local, build works successfully.

This is a major issue for us !

**Comment by n4ss** [comment]
@goffinf did you open an issue in https://github.com/awslabs/amazon-ecr-credential-helper too?

can you make a `docker-credential-ecr-login list` and check that the protocol for `123456789012.dkr.ecr.eu-west-1.amazonaws.com` is `https`?

**Comment by mpalmer** [comment]
We're hitting this problem, too, using the ECR creds helper.  I've done a bunch of digging around, and this is not a bug in that, except insofar as the cred helper protocol is insufficient for the task.

What I think `docker build` should do is to use the list of registries in `credHelpers` to get the list of credentials to send, rather than asking each cred helper what they think is a list of valid registries.  But, given the never-ending PRs that may, one day, result in us not having to pre-populate every registry ever, I doubt such a PR would be accepted (the more intelligent one, [to parse the `Dockerfile` to find registries to send creds for](https://github.com/moby/moby/pull/32273), got closed in favour of another PR to do cred negotiation -- which also got closed), so I'm not going to take the time to put a PR together unless there's a clear indication from Docker core that it had a hope of being merged.

We're going to start doing an explicit pull before every build, which neatly avoids this problem, and has the added benefit of making sure we're always building against the latest version of a tag, rather than a potentially out-of-date version of the same tag (non-immutable tags ftl).
